Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/GoogleCloudDataProtectionBuilderExtensions.cs

```diff
@@ -27,34 +30,55 @@ public static class GoogleCloudDataProtectionBuilderExtensions
 
         /// <summary>
         /// Configures the data protection system to persist keys in Google Cloud Storage.
         /// </summary>
-        /// <param name="builder">The data protection builder to configure.</param>
-        /// <param name="client">The Google Cloud Storage client to use for network requests.</param>
-        /// <param name="bucketName">The name of the bucket in which to store the object containing the keys.</param>
-        /// <param name="objectName">The name of the object in which to store the keys.</param>
+        /// <remarks>
+        /// <para>
+        /// If <paramref name="client"/> is null, the client is constructed as follows:
+        /// <list type="bullet">
+        ///   <item><description>If a <c>StorageClient</c> is configured via dependency injection, that is used.</description></item>
+        ///   <item><description>If a <c>GoogleCredential</c> is configured via dependency injection, that is used to construct a <c>StorageClient</c>.</description></item>
+        ///   <item><description>Otherwise, the default credentials are used to construct a <c>StorageClient</c>.</description></item>
+        /// </list>
+        /// </para>
+        /// </remarks>
+        /// <param name="builder">The data protection builder to configure. Must not be null.</param>
+        /// <param name="bucketName">The name of the bucket in which to store the object containing the keys. Must not be null.</param>
+        /// <param name="objectName">The name of the object in which to store the keys. Must not be null.</param>
+        /// <param name="client">The Google Cloud Storage client to use for network requests. May be null, in which case the client will
+        /// be fetched from dependency injection or created with the default credentials.</param>
         /// <returns>The same builder, for chaining purposes.</returns>
         public static IDataProtectionBuilder PersistKeysToGoogleCloudStorage(
-            this IDataProtectionBuilder builder, StorageClient client, string bucketName, string objectName)
+            this IDataProtectionBuilder builder, string bucketName, string objectName, StorageClient client)
         {
-            builder.Services.Configure<KeyManagementOptions>(options =>
-                options.XmlRepository = new CloudStorageXmlRepository(client, bucketName, objectName));
+            GaxPreconditions.CheckNotNull(builder, nameof(builder));
+            GaxPreconditions.CheckNotNull(bucketName, nameof(bucketName));
+            GaxPreconditions.CheckNotNull(objectName, nameof(objectName));
+
+            builder.Services.AddSingleton<IConfigureOptions<KeyManagementOptions>>(services =>
+            {
+                if (client == null)
+                {
+                    client = services.GetService<StorageClient>();
+                    if (client == null)
+                    {
+                        var credential = services.GetService<GoogleCredential>();
```

Comment:
I think we need to have a chat about this.
This is different to how the `Google.Apis.Auth.AspNetCore` package deals with user login - see an example of use here: https://github.com/googleapis/google-api-dotnet-client/blob/master/Src/Support/Google.Apis.Auth.AspNetCore.IntegrationTests/Controllers/HomeController.cs#L60

Reply:
Definitely happy to have a chat. But note that this isn't intended to be a *user* credential.

Reply:
As discussed, it's unfortunate in this context that a generic `GoogleCredential` may be either a user or service credential. As this is in alpha it's OK, but I think this further discussion before GA.